plutus-use-cases/src/Plutus/Contracts/Uniswap.hs

```diff
@@ -0,0 +1,988 @@
+import           Ledger.Constraints.TxConstraints as Constraints
+import qualified Ledger.Typed.Scripts             as Scripts
+import           Ledger.Value                     as Value
+import           Playground.Contract
+import           Plutus.Contract                  hiding (when)
+import qualified Plutus.Contracts.Currency        as Currency
+import qualified PlutusTx
+import           PlutusTx.Prelude                 hiding (Semigroup (..), unless)
+import           Prelude                          (Semigroup (..))
+import qualified Prelude
+import           Text.Printf                      (printf)
+
+feeNum, feeDen :: Integer
+feeNum = 3
+feeDen = 1000
+
+uniswapTokenName, poolStateTokenName :: TokenName
+uniswapTokenName = "Uniswap"
+poolStateTokenName = "Pool State"
+
+-- | A pair consisting of a 'CurrencySymbol' and a 'TokenName'.
+-- Coins are the entities that can be swapped in the exchange.
+data Coin = Coin
+    { cCurrency :: CurrencySymbol
+    , cToken    :: TokenName
+    } deriving (Show, Generic, ToJSON, FromJSON, ToSchema)
+
+PlutusTx.unstableMakeIsData ''Coin
+PlutusTx.makeLift ''Coin
+
+instance Eq Coin where
+    {-# INLINABLE (==) #-}
+    c == d = cCurrency c == cCurrency d && cToken c == cToken d
+
+instance Prelude.Eq Coin where
+    (==) = (==)
+
+{-# INLINABLE compareCoins #-}
+compareCoins :: Coin -> Coin -> Ordering
+compareCoins c d = case compare (cCurrency c) (cCurrency d) of
+    LT -> LT
+    GT -> GT
+    EQ -> compare (cToken c) (cToken d)
+
+instance Prelude.Ord Coin where
+    compare = compareCoins
+
+{-# INLINABLE coinLT #-}
+coinLT :: Coin -> Coin -> Bool
+coinLT c d = case compareCoins c d of
+    LT -> True
+    _  -> False
+
+{-# INLINABLE coin #-}
+-- | @'coin' c n@ denotes the value given by @n@ units of @'Coin'@ @c@.
+coin :: Coin    -- ^ The 'Coin'.
+     -> Integer -- ^ The desired number coins.
+     -> Value   -- ^ The 'Value' consisting of the given number of units of the given 'Coin'.
+coin Coin{..} = Value.singleton cCurrency cToken
+
+{-# INLINABLE coinValueOf #-}
+-- | Calculates how many units of the specified 'Coin' are contained in the
+-- given 'Value'.
+coinValueOf :: Value   -- ^ The 'Value' to inspect.
+            -> Coin    -- ^ The 'Coin' to look for.
+            -> Integer -- ^ The number of units of the given 'Coin' contained in the given 'Value'.
+coinValueOf v Coin{..} = valueOf v cCurrency cToken
+
+{-# INLINABLE hashCoin #-}
+hashCoin :: Coin -> ByteString
+hashCoin Coin{..} = sha2_256 $ concatenate (unCurrencySymbol cCurrency) (unTokenName cToken)
+
+data Sqrt =
+      Imaginary
+    | Exact Integer
+    | Irrational Integer
+    deriving stock Show
+
+PlutusTx.unstableMakeIsData ''Sqrt
+PlutusTx.makeLift ''Sqrt
+
+{-# INLINABLE rsqrt #-}
+rsqrt :: Integer -> Integer -> Sqrt
+rsqrt n d
+    | n * d < 0 = Imaginary
+    | n == 0    = Exact 0
+    | n == d    = Exact 1
+    | n < 0     = rsqrt (negate n) (negate d)
+    | otherwise = go 1 $ 1 + divide n d
```

Comment:
pretty sure this will be `error` if `d == 0`. Kind of obvious, but perhaps worth stating in the doc.